python/cudf/cudf/core/series.py

```diff
@@ -677,25 +633,34 @@ def __init__(
                 )
                 is property
             )
-            data = column.as_column(
+            column = as_column(
                 data,
                 nan_as_null=nan_as_null,
                 dtype=dtype,
                 length=len(index) if index is not None else None,
             )
             if copy and has_cai:
-                data = data.copy(deep=True)
-        else:
-            if dtype is not None:
-                data = data.astype(dtype)
+                column = column.copy(deep=True)
 
-        if index is not None and not isinstance(index, BaseIndex):
-            index = as_index(index)
+        assert isinstance(column, ColumnBase)
 
-        assert isinstance(data, ColumnBase)
+        if dtype is not None:
+            column = column.astype(dtype)
 
-        super().__init__({name: data})
-        self._index = RangeIndex(len(data)) if index is None else index
+        if index is not None:
+            index = as_index(index)
+        else:
+            index = RangeIndex(len(column))
+
+        if name_from_data is not None and name is None:
+            name = name_from_data
+        super().__init__({name: column}, index=index)
+        if index_from_data is not None:
+            # TODO: This there a better way to do this?
+            index_from_data = as_index(index_from_data)
+            reindexed = self.reindex(index=index_from_data, copy=False)
```

Comment:
There seems to be a `_reindex` function that can take `inplace=True` parameter.

Reply:
I just tried this and there were some test failures around data types not being preserved during the `_reindex`. I guess I'll just do this for now